packages/core-utils/src/nav/nav-utils.ts

```diff
@@ -0,0 +1,36 @@
+interface KeyedNavFn<Navs, FnReturn> {
+  <K extends keyof Navs>(key: K, args: Navs[K]): FnReturn;
+}
+
+/**
+ * Provides type safety for decoupled navigation mechanics.
+ *
+ * navTo and getUrl (key, args) functions are typed based on NavTypes generic to
+ * narrow args based on the type of NavTypes[key] contract.  NavTypes can hold a
+ * number of keyed arg contracts.
+ *
+ * @example
+ * interface MyNavActions {
+ *   key1: { a: string, b: number },
+ *   key2: null
+ * }
+ *
+ * // using NavLinkProvider<MyNavActions> type will have getUrl narrow
+ * // its 2nd argument based on the first argument:
+ *
+ * myProvider.getUrl('key1', args: {a: string, b: number}) => string
+ *
+ * // and similar for navTo
+ */
+export interface NavLinkProvider<NavTypes> {
```

Comment:
Maybe "NavProvider" instead of "NavLinkProvider". Since there's no longer a `getLink` method, the "link" in the name here seems out of place.

Reply:
I avoided NavProvider since I am thinking that will be a broader interface with additional app-level utils like we see in Terra's nav.ts module.  I could go for NavActionProvider maybe?

Reply:
Eh, but then `getUrl` isn't really an action. I'm not entirely convinced there's a conflict in names: I could see an ApplicationNavProvider that extends the basic NavProvider interface presented to components.

Reply:
I am guided by the I in SOLID (Interface Segregation) which says to prefer narrow interfaces over broad ones.  This is a narrow interface just for nav-link functionality... so I feel a more specific name makes sense.  NavLinkProvider seems a good fit following that reasoning.  NavProvider is a broader, less specific name.

Reply:
"Link" makes me think `<a href=...`. This is broader than that, since it allows navigation from any code, not only links, via the `navTo` function.